kong/dns/client.lua

```diff
@@ -0,0 +1,691 @@
+    end
+  end
+
+  log(NOTICE, PREFIX, "supported types: ", enable_srv and "srv " or "",
+              enable_ipv4 and "ipv4 " or "", enable_ipv6 and "ipv6 " or "")
+
+  -- parse resolv.conf
+  local resolv, err = parse_resolv_conf(opts.resolv_conf, opts.enable_ipv6)
+  if not resolv then
+    log(WARN, PREFIX, "Invalid resolv.conf: ", err)
+    resolv = { options = {} }
+  end
+
+  -- init the resolver options for lua-resty-dns
+  local nameservers = (opts.nameservers and not table_isempty(opts.nameservers))
+                      and opts.nameservers
+                      or resolv.nameservers
+
+  if not nameservers or table_isempty(nameservers) then
+    log(WARN, PREFIX, "Invalid configuration, no nameservers specified")
+  end
+
+  local r_opts = {
+    retrans = opts.retrans or resolv.options.attempts or 5,
+    timeout = opts.timeout or resolv.options.timeout or 2000, -- ms
+    no_random = opts.no_random or not resolv.options.rotate,
+    nameservers = nameservers,
+  }
+
+  -- init the mlcache
+
+  -- maximum timeout for the underlying r:query() operation to complete
+  -- socket timeout * retrans * 2 calls for send and receive + 1s extra delay
+  local lock_timeout = r_opts.timeout / 1000 * r_opts.retrans * 2 + 1 -- s
+
+  local resty_lock_opts = {
+    timeout = lock_timeout,
+    exptimeout = lock_timeout + 1,
+  }
+
+  -- TODO: convert the ipc a module constant, currently we need to use the
+  --       ipc_source to distinguish sources of different DNS client events.
+  ipc_counter = ipc_counter + 1
+  local ipc_source = "dns_client_mlcache#" .. ipc_counter
+  local ipc = {
+    register_listeners = function(events)
+      -- The DNS client library will be required in globalpatches before Kong
+      -- initializes worker_events.
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local cwid = worker_id() or -1
+      for _, ev in pairs(events) do
+        local handler = function(data, event, source, wid)
+          if cwid ~= wid then -- Current worker has handled this event.
+            ev.handler(data)
+          end
+        end
+
+        kong.worker_events.register(handler, ipc_source, ev.channel)
+      end
+    end,
+
+    -- @channel: event channel name, such as "mlcache:invalidate:dns_cache"
+    -- @data: mlcache's key name, such as "<qname>:<qtype>"
+    broadcast = function(channel, data)
+      if not kong or not kong.worker_events then
+        return
+      end
+
+      local ok, err = kong.worker_events.post(ipc_source, channel, data)
+      if not ok then
+        log(ERR, PREFIX, "failed to post event '", ipc_source, "', '", channel, "': ", err)
+      end
+    end,
+  }
+
+  local cache, err = mlcache.new("dns_cache", "kong_dns_cache", {
+    ipc = ipc,
+    neg_ttl = opts.error_ttl or DEFAULT_ERROR_TTL,
+    -- 10000 is a reliable and tested value from the original library.
+    lru_size = opts.cache_size or 10000,
+    shm_locks = ngx.shared.kong_locks and "kong_locks",
```

Comment:
In what case might we not have `ngx.shared.kong_locks` and still enable the new DNS client?

Reply:
The initial resty command of `bin/kong` has no lua_shared_dict loaded